File: problem-solving-patterns/sliding-window-pattern/slidingWindowPattern.ts
// O(n) time but with O(n) space
export function maxSubarraySum(numbers: number[], n: number) {
  if (numbers.length < n) return null;

  let max = -Infinity;

  const lastSubarrayStartIndex = numbers.length - n;

  for (let i = 0; i <= lastSubarrayStartIndex; i++) {
    const subarray = numbers.slice(i, i + n);
    const sum = sumArray(subarray);
    if (sum > max) {
      max = sum;
    }
  }
  return max;
}

//  0, 1, 2, 3, 4 (index)
// [1, 2, 3, 4, 5], n = 3
//

function sumArray(numbers: number[]) {
  return numbers.reduce((acc, val) => acc + val);
}

export function maxSubarraySumConstantSpace(
  numbers: number[],
  subarraySize: number
) {
  if (numbers.length < subarraySize) return null;
  let currentSum = 0;

  // loop once to sum initial window
  for (let i = 0; i < subarraySize; i++) {
    currentSum += numbers[i];
  }
  let max = currentSum;

  for (let i = 1; i < numbers.length; i++) {
    // slide the window
    const removedNumber = numbers[i - 1];
    const addedNumber = numbers[i + subarraySize - 1];
    currentSum = currentSum - removedNumber + addedNumber;

    if (currentSum > max) {
      max = currentSum;
    }
    // could also use max = Math.max(max, currentSum)
  }
  return max;
}

export function minSubarrayLen(arr: number[], minSum: number) {
  let subarrayLength = 1;
  let largestNumberIndex = 0;
  // Find the max value to decide where to grow subarray from
  for (let i = 0; i < arr.length; i++) {
    if (arr[i] > arr[largestNumberIndex]) {
      largestNumberIndex = i;
    }
  }
  // Loop only until subarraySum is smaller than min sum and also,
  // subarray length is smaller than array length
  let subarraySum = arr[largestNumberIndex];
  let subarrayHeadIndex = largestNumberIndex;
  let subarrayTailIndex = largestNumberIndex;
  while (subarraySum < minSum && subarrayLength < arr.length) {
    if (shouldGrowLeft(arr, subarrayHeadIndex, subarrayTailIndex)) {
      subarrayHeadIndex--;
      subarraySum += arr[subarrayHeadIndex];
    } else {
      subarrayTailIndex++;
      subarraySum += arr[subarrayTailIndex];
    }
    subarrayLength++;
  }
  if (subarraySum >= minSum) return subarrayLength;
  return 0;
}

function shouldGrowLeft(
  arr: number[],
  subarrayHeadIndex: number,
  subarrayTailIndex: number
) {
  // handle edge cases
  if (subarrayHeadIndex === 0 && subarrayTailIndex === arr.length - 1)
    throw Error('Subarray cannot grow anymore');
  if (subarrayHeadIndex === 0) return false;
  if (subarrayTailIndex === arr.length - 1) return true;
  // look backward and ahead to decide in which direction subarray should grow
  const beforeValue = arr[subarrayHeadIndex - 1];
  const afterValue = arr[subarrayTailIndex + 1];
  if (beforeValue > afterValue) return true;

  return false;
}

export function minSubArrayLenAnswer(numbers: number[], minSum: number) {
  let sum = 0;
  let headIndex = 0;
  let tailIndex = 0;
  let minLen = Infinity;

  while (headIndex < numbers.length) {
    // Update sum and move the tail if sum < minSum and tail is in range
    if (sum < minSum && tailIndex < numbers.length) {
      // when tail reached the end, last value will be added to sum,
      // and the tail is pushed out of the numbers array
      sum += numbers[tailIndex];
      tailIndex++;
    }
    // If the sum became geq to minSum && currentWindowLength < minLen, update minLen
    // Also move the head resetting the window and current sum
    else if (sum >= minSum) {
      // take one off since tail was moved after getting the sum
      const currentWindowLength = tailIndex - headIndex + 1 - 1;
      minLen = Math.min(minLen, currentWindowLength);

      sum -= numbers[headIndex];
      headIndex++;
    }
    // Tail is moved out of the array
    else {
      break;
    }
  }
  // Tail or Head is moved out of the array
  return minLen === Infinity ? 0 : minLen;
}

export function findLongestSubstring(str: string) {
  if (str.length <= 1) return str.length;
  // str.length >= 2
  let maxLen = 1;
  let headIndex = 0;
  let tailIndex = 1;
  const headValue = str[headIndex];
  let chars = { [headValue]: headIndex };
  // loop while indices are within the range
  while (headIndex < str.length && tailIndex < str.length) {
    const tailValue = str[tailIndex];
    // tail value is unique WITHIN the window
    const isTailValueUnique =
      chars[tailValue] === undefined || chars[tailValue] < headIndex;
    if (isTailValueUnique) {
      // register the new unique value and save its index
      chars[tailValue] = tailIndex;
    } else {
      // set the head right after the old existing char
      headIndex = chars[tailValue] + 1;
      // update the char index with the index of the duplicated value
      // so that we can keep the index within the window
      chars[tailValue] = tailIndex;
    }
    // update the max length with the current head & tail index
    maxLen = Math.max(maxLen, tailIndex - headIndex + 1);

    tailIndex++;
  }
  return maxLen;
}
// "abcbac"
//    h
//     t

// bbb
// h
//  t

export function findLongestSubstringWithForLoop(str: string) {
  let longest = 0;
  let headIndex = 0;
  const seen = {} as { [char: string]: number };
  // Find all substring with non-repeating characters and keep records of the longest
  for (let tailIndex = 0; tailIndex < str.length; tailIndex++) {
    const tailValue = str[tailIndex];
    const lastSeenIndex = seen[tailValue];
    if (lastSeenIndex !== undefined) {
      // headIndex can't go back
      headIndex = Math.max(lastSeenIndex + 1, headIndex);
    }
    seen[tailValue] = tailIndex;
    const newSubstringLength = tailIndex - headIndex + 1;
    longest = Math.max(longest, newSubstringLength);
  }
  return longest;
}

// thecatinthehat
//       h
//          t

File: problem-solving-patterns/sliding-window-pattern/sliding_window_pattern.py
from typing import List, Union


def max_subarray_sum(numbers: List[int], size: int) -> Union[int, None]:
    """Calculates the maximum sum of the subarray of the given size

    Args:
        numbers: a list of integers
        size: the size fo subarray

    Returns:
        int: the sum of the subarray. None if the array is empty
    """
    if size > len(numbers):
        return None
    current_sum = sum(numbers[0:size])
    max_sum = current_sum
    for i in range(1, len(numbers) - size):
        removing_number = numbers[i - 1]
        adding_number = numbers[i + size - 1]
        current_sum = current_sum - removing_number + adding_number
        max_sum = max(max_sum, current_sum)

    return max_sum


def where_to_grow(numbers: List[int], head_index: int, tail_index: int) -> int:
    try:
        left = numbers[head_index - 1]
    except IndexError:
        left = -1
    try:
        right = numbers[tail_index + 1]
    except IndexError:
        right = -1
    if left < 0 and right < 0:
        return 0
    elif left < 0 or left < right:
        return 1
    else:
        return -1


def min_subarray_len(numbers: List[int], min: int) -> int:
    # Invalidate and return early!
    if sum(numbers) < min:
        return 0
    subarray_len = 0
    subarray_sum = 0
    try:
        max_number = max(numbers)
    except ValueError:  # given an empty list
        return 0
    max_index = numbers.index(max_number)
    subarray_sum += max_number
    subarray_len += 1
    subarray_head_index = max_index
    subarray_tail_index = max_index
    while subarray_head_index >= 0 and subarray_tail_index < len(numbers):
        if subarray_sum >= min:
            return subarray_len
        direction = where_to_grow(numbers, subarray_head_index, subarray_tail_index)
        if direction == -1:
            subarray_head_index -= 1
            subarray_sum += numbers[subarray_head_index]
            subarray_len += 1
        elif direction == 1:
            subarray_tail_index += 1
            subarray_sum += numbers[subarray_tail_index]
            subarray_len += 1
        else:
            return 0

    return subarray_len


def find_longest_substring(string: str) -> int:
    substring_head_index = 0
    longest_substring_length = 0
    char_dict = {}

    for i, char in enumerate(string):

        def update_longest_substring_length():
            nonlocal longest_substring_length
            current_substring_length = i - substring_head_index + 1
            longest_substring_length = (
                current_substring_length
                if current_substring_length > longest_substring_length
                else longest_substring_length
            )

        # the character exists in substring
        try:
            prev_char_index = char_dict[char]
            substring_includes_char = prev_char_index >= substring_head_index
            if substring_includes_char:
                # move substring head forward to exclude old duplicated character
                substring_head_index = char_dict[char] + 1
                # update the index of the duplicated character with the index of the new one
                char_dict[char] = i
            update_longest_substring_length()

        # character not found in the dictionary
        except KeyError:
            char_dict[char] = i
            update_longest_substring_length()

    return longest_substring_length


#
# thisisawesome
#     h
#      i
# char_dict : { t: 0, h: 1, i: 4, s: 5}
# longest_substring_length: 1
